File: InterviewBit/L6-Tree_Data_Structure/inorder-traversal.cpp
#include <iostream>
#include <stack>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

vector<int> inorderTraversal(TreeNode *A) {
  vector<int> in;
  if (!A)
    return in;
  stack<TreeNode *> s;
  while (true) {
    while (A) {
      s.push(A);
      A = A->left;
    }
    if (s.empty())
      break;
    A = s.top();
    s.pop();
    in.push_back(A->val);
    A = A->right;
  }
  return in;
}

int main(int argc, char const *argv[]) {
  /* code */
  return 0;
}

File: 20210816-0947/OOD_x_LLD/src/InterviewBit/L6_Tree_Data_Structure/InorderTraversal.java
package InterviewBit.L6_Tree_Data_Structure;

import java.util.ArrayList;
import java.util.Stack;

public class InorderTraversal {

    public int[] inorderTraversal(TreeNode A) {
        if (A == null) return null;
        Stack<TreeNode> stack = new Stack<>();
        ArrayList<Integer> output = new ArrayList<>();
        TreeNode root = A;
        while (true) {
            while (root != null) {
                stack.push(root);
                root = root.left;
            }
            if (stack.empty()) break;
            root = stack.peek();
            stack.pop();
            output.add(root.val);
            root = root.right;
        }
        int[] result = new int[output.size()];
        for (int i = 0; i < output.size(); i++) {
            result[i] = output.get(i);
        }
        return result;
    }
}
